repoupdater/github: Fetch GitHub repositories in batches

mrnugget opened the pull request:
This PR adds a new method to the GitHub client, `GetRepositoriesByNameWithOwnerFromAPI`, which allows the caller to fetch repositories in batches, using the GraphQL API. It should fix https://github.com/sourcegraph/sourcegraph/issues/3907 but...

**IMPORTANT**: GitHub's GraphQL API only offers the fetching of repositories in batches by using GraphQL node IDs. But we don't have the node IDs at hand when doing our first sync, we only have a list of `"owner/repository-name"`s. That means we can't use the existing [getRepositoryByNodeIDFromAPI](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/pkg/extsvc/github/repos.go#L301)

Instead I decided to build GraphQL queries on the fly, using aliases to fetch multiple repositories in the same query. Here's an example query:

```graphql
fragment RepositoryFields on Repository {
  id
  nameWithOwner
  description
  url
  isPrivate
  isFork
  isArchived
}

{
  repo_sourcegraph_repository_1: repository(owner: "sourcegraph", name: "repository-1") {
    ... on Repository {
      ...RepositoryFields
    }
  }
  repo_sourcegraph_repository_2: repository(owner: "sourcegraph", name: "repository-2") {
    ... on Repository {
      ...RepositoryFields
    }
  }
  repo_sourcegraph_repository_3: repository(owner: "sourcegraph", name: "repository-3") {
    ... on Repository {
      ...RepositoryFields
    }
  }
}
```

Of course I tried to find out how many repositories I could query in one request like this, but neither the GitHub API documentation nor the GraphQL spec were any help here (pointers appreciated!). So I did some experimentation and as it turns out, **the API stops responding with results when specifying more than 37 repositories in one query**.

Since 37 is a weird number I rounded it down to 30 and used that as the batch size.

But still: since I'm a GraphQL newbie I feel like I'm relying on undefined behavior and I'm not sure whether the GitHub API will keep it up.

Please see this PR as an idea/proposal and cause for discussion: I would love to know what you think about using the API like this and what I'm possibly missing.

Test plan: `go test` and manual testing with an external service configuration that has 100 repos in

tsenart commented on pkg/extsvc/github/repos.go:
```diff
@@ -379,6 +382,74 @@ query Repositories($ids: [ID!]!) {
+	var result map[string]*Repository
+	err = c.requestGraphQL(ctx, token, query, map[string]interface{}{}, &result)
+	if err != nil {
+		if gqlErrs, ok := err.(graphqlErrors); ok {
```
This could be simplified like this:

```go
errs, _ := err.(graphqlErrors)
for _, e := range errs {
    if e.Type != graphqlErrTypeNotFound {
        return nil, e // I changed this from err too e. Did you mean to return the parent err?
    }
}
return nil, err
```

mrnugget replied:
Thanks! I adopted the code from the other methods. Question is: do we want to return an error if _one_ repository was not found or not? I'd say it makes sense to ignore a missing repo, because we don't want the syncer to fail when someone deletes/renames a GitHub repository without telling the Sourcegraph admin before.

tsenart commented on pkg/extsvc/github/repos.go:
```diff
@@ -379,6 +382,74 @@ query Repositories($ids: [ID!]!) {
+		}
+	}
+
+	repos := []*Repository{}
```
Declaring empty slices, usually, benefits from avoiding the eager allocation this incurs by using the `var repos []*Repository` syntax. See: https://github.com/golang/go/wiki/CodeReviewComments#declaring-empty-slices

tsenart replied:
In this case, however, we already know the upper bound of the repos we're working with, so this would be my proposal:

```suggestion
	repos := make([]*Repository, 0, len(result))
```

tsenart commented on pkg/extsvc/github/repos_test.go:
```diff
@@ -527,3 +527,118 @@ func TestClient_GetRepositoryByNodeID_security(t *testing.T) {
+		t.Fatal(err)
+	}
+
+	fmt.Printf("query=%q", query)
```
Leftover.

tsenart commented on pkg/extsvc/github/repos_test.go:
```diff
@@ -527,3 +527,118 @@ func TestClient_GetRepositoryByNodeID_security(t *testing.T) {
 	}
 }
+
+func TestClient_buildGetRepositoriesBatchQuery(t *testing.T) {
```
Is there something this unit test is meant to cover that a `TestClient_GetRepositoriesByNameWithOwnerFromAPI` doesn't?

mrnugget replied:
Yes, I specifically wanted to test what the query looks like and that it sanitizes the repo names. The test for the outer method doesn't cover this, since it uses a mocked response, which means you can construct an empty `query {}` and the test would still pass.

tsenart commented on pkg/extsvc/github/repos_test.go:
```diff
@@ -527,3 +527,118 @@ func TestClient_GetRepositoryByNodeID_security(t *testing.T) {
+			IsArchived:       true,
+			ViewerPermission: "ADMIN",
+		},
+	}
```
What do you think about testing some error cases too like `not found`?

mrnugget replied:
Good idea! Thanks for reminding me. I just wrote a test case and immediately found out that it does not work as expected :)

tsenart replied:
Since we introduced fall-back logic, do you think it's worth testing it here too?

mrnugget replied:
Since the fall-back logic resides in `repo-updater`, we'd have to test it there and I refrained from doing that because there were no pre-existing tests for the complete method

tsenart replied:
> pre-existing tests for the complete method

Would it be too onerous to cover that in here? https://github.com/sourcegraph/sourcegraph/blob/master/cmd/repo-updater/repos/sources_test.go#L128

tsenart commented:
> Please see this PR as an idea/proposal and cause for discussion: I would love to know what you think about using the API like this and what I'm possibly missing.

I think the idea makes total sense! But we should indeed try to figure out if this is undefined behaviour prone to change or not. If we can't gain such confidence, we could build a fallback code path that fetches repos one by one like before, in case the GraphQL API starts acting up.

mrnugget commented on cmd/repo-updater/repos/github.go:
```diff
@@ -310,30 +310,34 @@ func (s *GithubSource) listAllRepositories(ctx context.Context) ([]*github.Repos
-	for _, nameWithOwner := range s.config.Repos {
+	var (
+		batchSize = 30
+		lenRepos  = len(s.config.Repos)
```
Agree!

mrnugget commented on pkg/extsvc/github/repos.go:
```diff
@@ -379,6 +382,74 @@ query Repositories($ids: [ID!]!) {
+// the conservative step back from 37.
+//
+// This method does not cache.
+func (c *Client) GetRepositoriesByNameWithOwnerFromAPI(ctx context.Context, token string, namesWithOwners []string) ([]*Repository, error) {
```
I'd happily change the name and the API, but just FYI: I modelled this after the existing method above:

https://github.com/sourcegraph/sourcegraph/blob/abdf9ad3856483e525949d94acc1a3e74936ea41/pkg/extsvc/github/repos.go#L346

As I understand it the token is configurable for that method because it's changed on which user account is driving the client.

Since we don't have that requirement (yet), let's drop it :)


> Another thought: Have we tested this with Github Enterprise?

No, but will do!

mrnugget commented on cmd/repo-updater/repos/github.go:
```diff
@@ -310,30 +310,34 @@ func (s *GithubSource) listAllRepositories(ctx context.Context) ([]*github.Repos
+		for _, r := range repos {
+			set[r.DatabaseID] = r
+		}
 		time.Sleep(s.client.RateLimit.RecommendedWaitForBackgroundOp(1)) // 0-duration sleep unless nearing rate limit exhaustion
```
It does. The cost for one request is `1`.

mrnugget commented:
This just in: sometimes the sync with this new method fails because the GraphQL API returns an EOF?

```
15:18:31           github-proxy | 2019/05/27 15:18:31 Post https://api.github.com/graphql: EOF
15:18:31           repo-updater | WARN Failed to decode error response from github API, error: invalid character 'P' looking for beginning of value
15:18:31           repo-updater | ERROR source.list-repos, error: 1 error occurred:
15:18:31           repo-updater |       * Error getting GitHub repositories: [sourcegraph/sourcegraph-archive-private-20160601 sourcegraph/go-selenium sourcegraph/go-vcsurl sourcegraph/go-vcs sourcegraph/javascript-grapher sourcegraph/srcscan sourcegraph/httpcache sourcegraph/es-scorer sourcegraph/grapher-tutorial sourcegraph/clojure-grapher sourcegraph/programming-challenge sourcegraph/annotate sourcegraph/sourcegraph-sublime-old sourcegraph/makex sourcegraph/pydep sourcegraph/vcsstore sourcegraph/conf_private sourcegraph/talks sourcegraph/srclib-template-project sourcegraph/srclib-dummy sourcegraph/srclib sourcegraph/go-sourcegraph sourcegraph/srclib-go sourcegraph/srclib-python sourcegraph/srclib-ruby sourcegraph/srclib-javascript sourcegraph/srclib-java sourcegraph/srclib-sample sourcegraph/thesrc sourcegraph/sourcegraph-mode]: request to http://127.0.0.1:3180/graphql returned status 500:
15:18:31           repo-updater | ERROR Syncer, error: syncer.sync.sourced: 1 error occurred:
15:18:31           repo-updater |       * Error getting GitHub repositories: [sourcegraph/sourcegraph-archive-private-20160601 sourcegraph/go-selenium sourcegraph/go-vcsurl sourcegraph/go-vcs sourcegraph/javascript-grapher sourcegraph/srcscan sourcegraph/httpcache sourcegraph/es-scorer sourcegraph/grapher-tutorial sourcegraph/clojure-grapher sourcegraph/programming-challenge sourcegraph/annotate sourcegraph/sourcegraph-sublime-old sourcegraph/makex sourcegraph/pydep sourcegraph/vcsstore sourcegraph/conf_private sourcegraph/talks sourcegraph/srclib-template-project sourcegraph/srclib-dummy sourcegraph/srclib sourcegraph/go-sourcegraph sourcegraph/srclib-go sourcegraph/srclib-python sourcegraph/srclib-ruby sourcegraph/srclib-javascript sourcegraph/srclib-java sourcegraph/srclib-sample sourcegraph/thesrc sourcegraph/sourcegraph-mode]: request to http://127.0.0.1:3180/graphql returned status 500:
```

I'll try to figure out why that happens.

mrnugget commented on pkg/extsvc/github/repos.go:
```diff
@@ -379,6 +382,74 @@ query Repositories($ids: [ID!]!) {
+			return "", err
+		}
+		ident := strings.Replace(fmt.Sprintf("repo_%s_%s", owner, name), "-", "_", -1)
+		ident = strings.Replace(ident, ".", "_", -1)
```
Sounds good! But I want your opinion on this: theoretically I could name the aliases in the GraphQL query `repo1`, `repo2`, ... `repoN` since we never use that information. I initially used the repo names to keep the query debuggable, but a) the `name` and `owner` are mentioned anyway and b) this approach comes with the added cost of this "sanitization". What do you about switching to `repo1`-`repoN`?

tsenart replied:
> What do you about switching to repo1-repoN?

Sounds better!

mrnugget replied:
Done!

mrnugget commented on cmd/repo-updater/repos/github.go:
```diff
@@ -310,30 +310,34 @@ func (s *GithubSource) listAllRepositories(ctx context.Context) ([]*github.Repos
-		log15.Debug("github sync: GetRepository", "repo", repo.NameWithOwner)
-		set[repo.DatabaseID] = repo
+		log15.Debug("github sync: GetRepositoriesByNameWithOwnerFromAPI", "repos", batch)
+		log15.Info("github sync: GetRepositoriesByNameWithOwnerFromAPI", "batchSize", len(batch))
```
Leftover 🙈

mrnugget commented:
@tsenart I incorporated all of your feedback and added a fallback mechanism, where the `GithubSource` switches back to sequential mode if the batch fetching failed for any reason. I added this after another round of research regarding "how many aliases does GraphQL allow in a query?" or "what does GitHub's docs say about this?" yielded no result (but I contacted GitHub support and asked them about a hard limit)

mrnugget commented:
Update regarding "does this work with GitHub enterprise?": Yes, it does. I just tested this locally with our GHE instance and sending ~300 "update external service" requests to my local Sourcegraph instance. No errors, no fallbacks — everything worked as expected.

I'm currently trying to find out whether the EOF error mentioned above is related to `github-proxy`. So far, I can't reproduce it. It didn't occur once while doing ~600+ syncs, but it popped up twice when restarting my local instance, which leads me to suspect that it has something to do with `github-proxy` and sourcegraph internal connection management etc.

tsenart commented on cmd/repo-updater/repos/github.go:
```diff
@@ -361,6 +366,37 @@ func (s *GithubSource) getRepository(ctx context.Context, nameWithOwner string)
+		if err != nil {
+			return err
+		}
+		log15.Debug("github sync: GetGetReposByNameWithOwner", "repos", batch)
```
```suggestion

		log15.Debug("github sync: GetGetReposByNameWithOwner", "repos", batch)
```

tsenart commented on cmd/repo-updater/repos/github.go:
```diff
@@ -310,31 +310,36 @@ func (s *GithubSource) listAllRepositories(ctx context.Context) ([]*github.Repos
+				errs = multierror.Append(errs, errors.New("Invalid GitHub repository: nameWithOwner="+nameWithOwner))
+				break
+			}
+			repo, err := s.client.GetRepository(ctx, owner, name)
```
```suggestion

			repo, err := s.client.GetRepository(ctx, owner, name)
```

tsenart commented on cmd/repo-updater/repos/github.go:
```diff
@@ -310,31 +310,36 @@ func (s *GithubSource) listAllRepositories(ctx context.Context) ([]*github.Repos
+				}
+				errs = multierror.Append(errs, errors.Wrapf(err, "Error getting GitHub repository: nameWithOwner=%s", nameWithOwner))
+				break
+			}
```
```suggestion
			}

```

tsenart commented on cmd/repo-updater/repos/github.go:
```diff
@@ -310,31 +310,36 @@ func (s *GithubSource) listAllRepositories(ctx context.Context) ([]*github.Repos
+				break
+			}
+			log15.Debug("github sync: GetRepository", "repo", repo.NameWithOwner)
+			set[repo.DatabaseID] = repo
```
```suggestion
			set[repo.DatabaseID] = repo

```

tsenart commented on cmd/repo-updater/repos/github.go:
```diff
@@ -361,6 +366,37 @@ func (s *GithubSource) getRepository(ctx context.Context, nameWithOwner string)
+		log15.Debug("github sync: GetGetReposByNameWithOwner", "repos", batch)
+		for _, r := range repos {
+			set[r.DatabaseID] = r
+		}
```
```suggestion
		}

```

tsenart commented on pkg/extsvc/github/repos.go:
```diff
@@ -379,6 +383,76 @@ query Repositories($ids: [ID!]!) {
+		if err != nil {
+			return "", err
+		}
+		b.WriteString(fmt.Sprintf("repo%d: repository(owner: %q, name: %q) { ", i, owner, name))
```
```suggestion
		fmt.Fprintf(&b, "repo%d: repository(owner: %q, name: %q) { ", i, owner, name))
```

tsenart replied:
Actually, you can coalesce these two calls into one:

```go
fmt.Fprintf(&b,`repo%d: repository(owner: %q, name: %q) {
  ... on Repository { ...RepositoryFields }
}`, i, owner, name)
```

tsenart commented:
> I'm currently trying to find out whether the EOF error mentioned above is related to github-proxy. So far, I can't reproduce it.

@keegancsmith: Can we just get rid of `github-proxy`?

tsenart commented on cmd/repo-updater/repos/github_test.go:
```diff
@@ -12,3 +26,104 @@ func TestExampleRepositoryQuerySplit(t *testing.T) {
+		t.Helper()
+
+		if have, want := len(rs), 2; have != want {
+			t.Fatalf("wrong number of repos yielded. want=%d, have=%d", want, have)
```
This assertion is not necessary together with the one below.

mrnugget replied:
Ah, right, I can now get rid of that one, thanks!

mrnugget commented:
Two updates:

1. I've now added tests for the fallback logic, which gives us additional safety that it a) works and b) won't be removed by accident
2. I sent multiple hundred of `sync-external-service` requests to my local Sourcegraph instance to trigger syncs for Github.com and Github Enterprise and I couldn't reproduce the EOF error once.

I think it's safe to merge/ship these changes. Since we have the fallback in place, the worst that could happen is an additional request.
